### ISSUE TITLE
Scripts per language, and most likely language per script API?

### ISSUE THREAD

**Issue by behdad** [harfbuzz/harfbuzz#1288]
We use those in Pango.  Generally useful for itemization.  @dscorbett something you can work on?  Is there reliable data available for these in CLDR?

**Comment by brawer** [comment]
CLDR has the most likely script given a language: https://www.unicode.org/repos/cldr/trunk/common/supplemental/likelySubtags.xml

**Comment by dscorbett** [comment]
I can put CLDR’s relevant data into HarfBuzz, but first I’d like to understand the use case better and how it relates to itemization. An itemizer gets the language from a `lang` tag or equivalent or falls back to 'dflt'; when would it need to guess based on the script? Why do a language’s possible scripts matter, as opposed to the actual script of the text being shaped?

**Comment by Richard57** [comment]
Indic data in CLDR mapping language to script is unreliable; it contains such howlers as claiming that Northern Thai is normally written in the Lanna script and that the Saurashtra language is normally written in the Saurashtra script. I'm also sceptical as to runic being the dominant script for Old Norse.  The principal script of Old Irish is Latin, not Ogam, and, if Michael Everson is to be believed, the most frequent use of Ogam is for English (en_Ogam_IE), not Old (or even Primitive) Irish, though the numbers may shift if handwritten material is excluded.

Although my sample is limited and biased, what I've encountered in Runic texts a paragraph or more long is mostly en_Runr_UK ('The Hobbit' plus a colleague's whiteboard notes), with a bit of en_Runr_NZ ('The Almighty Johnsons').

In short, the CLDR data on guessing language from script or script from language is frequently unreliable.

**Comment by behdad** [comment]
I thought more about this...  Let me explain what Pango does today.  In the next comment I share my latest thinking.

Pango uses `$LANGUAGE` to fetch list of user's language.  Every system has its own way but in the end, user's preferred languages comes down to a simple list.  Eg. `en:fa` means "English, Persian" in that order.

Pango then needs to resolve the best language tag for each segment.  This is necessary because HarfBuzz needs that for shaping, and this is essential for choosing correct language-specific forms.  So for example if there's text that involves Latin, Arabic, and Thai scripts, this is what Pango does to those:

  - For Latin, it notices that the English language is written in Latin (using [`pango-script-lang-table.h`](https://gitlab.gnome.org/GNOME/pango/blob/master/pango/pango-script-lang-table.h), so it attaches language `en` to this segment.

  - For Arabic script, Pango rejects English being written in Arabic scripts; it looks at Persian next, and sees that Persian is written in Arabic script, so attaches language `fa` to this segment.

  - For Thai script, Pango finds no user language that is written in this script.  Failing to find a user language, Pango uses [`pango_script_get_sample_language`](https://gitlab.gnome.org/GNOME/pango/blob/master/pango/pango-language.c#L878) to mark this segment as Thai language.

When I filed this issue first, I wanted to provide relevant functions from HarfBuzz to enable implementing similar logic in other layout systems.  Since then, I've changed my mind and have a different plan.  Next commit.

**Comment by behdad** [comment]
Now noticed that if HarfBuzz added buffer API to accept a *list* of languages, instead of one language, then all the complex logic mentioned above can be implemented in HarfBuzz, instead of within every client.  So, we definitely should do that.

I also think the logic above can be simplified in the following ways if implemented in HarfBuzz

- We should NOT tag eg. Thai script text as Thai language if `th` was *not* in user languages.  Ie. we should accept a font's Script system's default LangSys in cases of scripts we don't have a preferred language for.  Forcing the script's most common language can result in undesired behavior if user has chosen fonts that default to other languages.

- We can simply pass the `en:fa` list to the OT tag matching logic.  We don't really need to attach one language to the segment.  So we pass `en:fa` to the OT langsys matching logic.  It looks in the font's `arab` script; it looks for `en` and doesn't find any langsys in `arab` script for `en`, so it then tries `fa`.

The above works fine.  The only time it can fail is if a font eg has `en` langsys in `arab` script system.  Ie. the font supports an esoteric "English written in Arabic script" locale (`en-Arab`?).

Currently, I suggest we implement the simple logic in this comment and ignore this problem until later.  If we want to make sure `en` in `arab` should *not* be used for `en` language, then we need to have a language-tag-to-assumed-script mapping.  But it's amazing how far we can go without any script vs language table whatsoever.

As for the API itself, I'm leaning towards *not* adding any language-list API and silently document and start accept colon-separated lists as `hb_language_t` as well.  Ie, `en:fa` is a language list consisting of two language tags `en` and `fa` in that order.

Thoughts?

cc @jfkthame @dscorbett @mhosken @drott

**Comment by mhosken** [comment]
tldr; I agree.

I think it's a bad idea to guess script from language, especially as the list can get pretty long. I have SIL's mapping list here: https://github.com/silnrsi/langtags/tree/master/pub and it starts getting pretty big. I think harfbuzz's automatic script analysis via codepoint script categories, is generally fine (Can't remember how well it supports script extensions).

With regard to language. I think language lists is a good idea. I'm not sure what I think about parsing the `$LANGUAGE` variable if no language list is passed. I think that should be an application driven choice: use default or use `$LANGUAGE` by whatever means. I would obviously say that I think that guessing language from script is a really bad idea, especially when fonts have the default langsys available precisely to support the situation where no language is specified (or the language(s) specified is not supported).

**Comment by behdad** [comment]
Thanks Martin.  I agree with you assessment as well; ie. I'm also not sure about picking up `$LANGUAGE` by default.  Though we *do* pick up `$LANG` and friends by default.

**Comment by drott** [comment]
In a browser / user-agent context, what would be pass as the language list? The values of the Accept-Language http header and what the user has configured to be an ordered list of supported languages? Or a preference we try to extract from the host system? E.g. the $LANGUAGES env var, or a Windows registry setting, or similar settings on other OSes?

**Comment by behdad** [comment]
> In a browser / user-agent context, what would be pass as the language list? The values of the Accept-Language http header and what the user has configured to be an ordered list of supported languages? Or a preference we try to extract from the host system? E.g. the $LANGUAGES env var, or a Windows registry setting, or similar settings on other OSes?

If CSS doesn't say anything about that, then I think the element's lang/xml:lang attribute followed by whatever ends up being sent as Accept-Language (which should come from user pref and system prefs i suppose).

**Comment by behdad** [comment]
Another API alternative is to keep `hb_buffer_set_language` as is, and add a `hb_buffer_set_fallback_languages` to take a list.  The singular language will be deemed as being at the head of the fallback list.  This way, clients wouldn't need to create a new list every time the current segment's language changes.

Maybe that's better approach?

If doing that way, still remains whether to accept `en:fa` as the `set_language` or do the abuse of `hb_language_t` at all or not.

**Comment by behdad** [comment]
Probably not gonna happen.
